core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -260,6 +261,64 @@ public UpdateSchema updateColumnDoc(String name, String doc) {
     return this;
   }
 
+  @Override
+  public UpdateSchema updateSchema(Schema newSchema) {
+    TypeUtil.visit(newSchema, new ApplyUpdates(schema, this));
+    return this;
+  }
+
+  private static final Joiner DOT = Joiner.on(".");
+
+  private static class ApplyUpdates extends TypeUtil.SchemaVisitor<Void> {
```

Comment:
If we cannot reliably distinguish between renames  and deletes + add. Maybe we should throw an error if we encounter any deletes when using this API?

Reply:
Since this uses field IDs, we can distinguish between a rename and a delete+add. We wouldn't be able to do that if we no longer have reliable field IDs though.